[package] Introduce BaseImporter to manage module namespace collisions.

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51976 [fx/package] make GraphModules packageable
* #51974 [fx] Separate globals assignment from code generation.
* **#51975 [package] Introduce BaseImporter to manage module namespace collisions.**
* #52320 [package] rename ex/importer.py to package_ex/importer.py

See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

suo committed:
[package] Introduce ModuleEnv to manage module namespace collisions.

See comments in code.

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

zdevito commented on torch/package/module_environment.py:
```diff
@@ -0,0 +1,219 @@
+            name: If set, use this name instead of looking up __name__ or __qualname__ on `obj`.
+                This is only here to match how Pickler handles __reduce__ functions that return a string,
+                don't use otherwise.
+            check_obj: If true, check that retrieving the returned name produces the same object as `obj`.
```
this flag feels weird. This should either always return None or always throw an exception if it cannot find the object.

I feel like the problem is that the error reporting for the mismatch is still happening in the pickler,  rather than just reporting the right error here: e.g. I found the object but it is not the same object directly in the exception.

suo replied:
Yeah, I was sort of struggling to not rewrite the pickler code too much. But maybe for the sake of this API not being silly it makes sense to be more intrusive (since we're editing the pickler code anyway…)

zdevito commented on torch/package/module_environment.py:
```diff
@@ -0,0 +1,219 @@
+            check_obj: If true, check that retrieving the returned name produces the same object as `obj`.
+                This should only be false if you plan to call `check_obj()` somewhere else.
+
+        Returns:
```
Reading more, I think we should drop check_obj as both a method and a flag here. This function should always throw ModuleEnvError if the objects do not match, and it should put into this error message the details about the object it did find if it doesn't match. I think pickler can be refactored around this, since it just immediately goes and calls check_obj.

zdevito commented on torch/package/module_environment.py:
```diff
@@ -0,0 +1,219 @@
+        """Returns true if this ModuleEnv only represents the default Python environment."""
+        return self._importers == [DefaultImporter]
+
+    def check_obj(self, obj, obj2, name=None):
```
This API is weird as well. It feels way to specialized to the pickler. I think all the functionality can be encapsulated in `get_name`.

zdevito commented on torch/package/module_environment.py:
```diff
@@ -0,0 +1,219 @@
+            name = obj.__name__
+
+        orig_module_name = self._whichmodule(obj, name)
+        obj_module_name = getattr(obj, "__module__", orig_module_name)
```
from here to the end of the function is an intertwined copy/paste. Can it be refactored into a function that returns `mangled, obj_location, obj_importer_name`

zdevito commented on torch/package/importer.py:
```diff
@@ -35,7 +45,7 @@ class PackageImporter:
     local to this importer.
     """
-    modules : Dict[str, Optional[types.ModuleType]]
+    modules : Dict[str, ModuleType]
```
This is wrong. `None` gets inserted if a module fails to load. Or it is allowed if you want to force a module to fail to load.

suo replied:
I looked a bit and could not find the place where we insert None, but possible I don't understand importlib. Can you point me to it?

zdevito commented on torch/package/exporter.py:
```diff
@@ -47,7 +48,7 @@ class PackageExporter:
     """
 
-    importers: List[Callable[[str], Any]]
+    importers: List[BaseImporter]
```
didn't you get rid of importers?

suo replied:
yes, oops

zdevito commented on test/test_package.py:
```diff
@@ -3,6 +3,7 @@
 from torch.testing._internal.common_utils import TestCase, run_tests, IS_WINDOWS
 from tempfile import NamedTemporaryFile
 from torch.package import PackageExporter, PackageImporter
+from torch.package.module_environment import ModuleEnv, DefaultImporter
```
If this is user facing it should be in `torch.package` and not just in the file it is defined in.

zdevito commented on test/test_package.py:
```diff
@@ -191,7 +192,7 @@ def test_save_imported_module_fails(self):
         f2 = self.temp()
         pe = PackageExporter(f2, verbose=False)
-        pe.importers.insert(0, importer1.import_module)
+        pe.module_env = ModuleEnv([importer1, DefaultImporter])
```
I don't like that users have to now know where to find DefaultImporter and if it is forgotten, then pickling is going to almost always fail (can't save things like builtins, etc.). This also doesn't let different parts of the exporter script register importers but requires that they are specified upfront when constructing ModuleEnv. Maybe we should keep `pe.importers` and just create a `ModuleEnv` in a @property on PackageExporter when requested.

suo replied:
I feel like it's good to be explicit and simple, even if a little verbose, given how strange the behavior is likely to be when things are wrong. Like, I feel like users should know that DefaultImporter is in there—if they forgot, their code is probably going to not work anyway (since doing `.append()` to an "empty" list is more natural than inserting at 0).

zdevito reviewed (commented):
I have a bunch of somewhat disparate minor comments to try to clean up the API.

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce ModuleEnv to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

zdevito commented on torch/package/importer.py:
```diff
@@ -0,0 +1,193 @@
+        return whichmodule(obj, name)
+
+
+SysImporter = _SysImporter()
```
`sys_importer` would be more pythonic because it is not a class.

zdevito replied:
Since it isn't a type, `sys_importer` is more pythonic.

zdevito commented on torch/package/package_exporter.py:
```diff
@@ -44,20 +44,16 @@ class PackageExporter:
-    When a name conflict occurs between importers, the first importer in the list takes precedence,
-    and only objects that refer to this first importers class can be saved
+    """A module environment that will be searched in order to find the module assocated with module names
+    referenced by other modules or by pickled objects. The default module environment just uses DefaultImporter,
```
Some old terminology here 'module environment' and `DefaultIMporter`

zdevito commented on torch/package/package_exporter.py:
```diff
@@ -44,20 +44,16 @@ class PackageExporter:
 
-
-    def __init__(self, f: Union[str, Path, BinaryIO], verbose: bool = True):
+    def __init__(self, f: Union[str, Path, BinaryIO], importer: Importer = SysImporter, verbose: bool = True):
```
It would be nice to allow `importer` to be `Either[Importer, List[Importer]]` and use it to initialize a OrderedImporter, since this will be a common case and prevent having to import the OrderedImporter object in user code.

zdevito commented on torch/package/importer.py:
```diff
@@ -0,0 +1,193 @@
+        """Find the module name an object belongs to.
+
+        This should be considered internal for end-users, but developers of
+        an importer can override it to customizer the behavior.
```
customize

zdevito commented on torch/package/importer.py:
```diff
@@ -0,0 +1,193 @@
+            importer_name = (
+                f"the importer for {get_mangle_prefix(module_name)}"
+                if is_mangled_
+                else "'DefaultImporter'"
```
DefaultImporter -> sys_importer

zdevito commented on torch/package/importer.py:
```diff
@@ -0,0 +1,193 @@
+            ObjNotFoundError: we couldn't retrieve `obj by name.
+            ObjMisMatchError: we found a different object with the same name as `obj`.
+        """
+        if name is None and obj and _Pickler.dispatch.get(type(obj)) is None:
```
Because this is the start of this function, and because it is only called in one place from our custom pickler, I think it  should go in that function and not here. I can see this accidentally firing for weird objects in other places.

zdevito replied:
Does that make sense or am I not understanding what the string version of reduce does?

suo replied:
afaict we need this here, because `__reduce__ -> str` is one way that objects can signal that they want to be global'd as a different name than what the pickler might infer. I ran across this when trying to package references to `typing.List` and friends in FX.

We can put equivalent logic in the import string formatter in `GraphModule`, but it made more sense to put it here and leave clients of the code unaware of this detail.

zdevito replied:
ok, I get it now. Given that definition of what it does, then it is in the right place.

zdevito commented on torch/package/importer.py:
```diff
@@ -0,0 +1,193 @@
+        return importlib.import_module(module_name)
+
+    def whichmodule(self, obj: Any, name: str) -> str:
+        return whichmodule(obj, name)
```
This looks like it will call the method and not the global. I guess it doesn't, but I'd feel better if we imported `whichmodule` under a different name.

zdevito commented on torch/package/importer.py:
```diff
@@ -0,0 +1,193 @@
+                return importer.get_name(obj, name)
+            except ObjNotFoundError as err:
+                last_err = err
+
```
we don't need it for this patch, but if you catch a ObjMismatchError here, then you can keep looking for the object, and if you succeed, you can report that the earlier importer shadowed the importer that you needed.

suo replied:
Yeah good idea. Right now we use the package mangling as an implementation detail to give a good message, but it would be better to do it without that knowledge.

zdevito reviewed (approved):
Cool! Just a couple minor things inline.

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]

suo committed:
Update on "[package] Introduce BaseImporter to manage module namespace collisions."


See comments in code.

Differential Revision: [D26340592](https://our.internmc.facebook.com/intern/diff/D26340592)

[ghstack-poisoned]
